crates/driver/src/infra/notify/mod.rs

```diff
@@ -0,0 +1,37 @@
+use {
+    super::Solver,
+    crate::domain::competition::{auction, solution},
+};
+
+mod notification;
+
+pub use notification::{Kind, Notification};
+
+pub fn empty_solution(solver: &Solver, auction_id: Option<auction::Id>) {
+    solver.notify(auction_id, notification::Kind::EmptySolution);
+}
+
+pub fn price_violation(solver: &Solver, auction_id: Option<auction::Id>) {
+    solver.notify(auction_id, notification::Kind::PriceViolation);
+}
+
+pub fn scoring_failed(solver: &Solver, auction_id: Option<auction::Id>) {
+    solver.notify(auction_id, notification::Kind::ScoringFailed);
+}
+
+pub fn encoding_failed(
```

Comment:
These helper functions feel weird to me.

On the one hand you already have `Solver::notify` and the concept of a `Notification`. Maybe `Solver::notify` could take a `Notification` which contains all the relevant information in each enum variant so you can drop this file?
On the other hand that would require you to move the pattern matching of `encoding_failed()` into the domain logic which doesn't seem right.
That solution would also have to have `Notification` in the `domain` to adhere to DDD which seems wrong as well.

Maybe this is just one of those cases were we have to live with boiler plate code for the sake of encapsulation.

This is not really a concrete suggestion for this PR just me trying to explore the DDD design philosophy using this PR as an example. 😅